app/services/katello/event_daemon.rb

```diff
@@ -1,60 +1,95 @@
-        return false
+      def pid
+        return unless pid_file && File.exist?(pid_file)
+
+        File.open(pid_file) { |f| f.read.to_i }
       end
 
-      if started?
-        Rails.logger.info("Katello Event Daemon already running; not attempting startup")
-        return false
+      def pid_file
+        settings[:pid_file]
       end
 
-      !::Foreman.in_rake? && !Rails.env.test? && SETTINGS[:katello][:event_daemon_enabled]
-    end
+      def write_pid_file
+        return unless pid_file
 
-    def self.start
-      if runnable?
-        services.values.each(&:run)
+        File.open(pid_file, 'w') { |f| f.puts Process.pid }
+      end
+
+      def stop
+        return unless pid == Process.pid
+        services.values.each(&:close)
+        File.unlink(pid_file) if pid_file && File.exist?(pid_file)
+      end
 
-        at_exit do
-          stop
+      def start(worker: false)
+        return if !runnable? || settings[:multiprocess] && !worker
+
+        lockfile = File.open(settings[:lock_file], 'r')
+        begin
+          lockfile.flock(File::LOCK_EX)
+          return if started? # ensure it wasn't started while we waited for the lock
+
+          start_services
+          write_pid_file
+
+          at_exit do
+            stop
+          end
+
+          Rails.logger.info("Katello event daemon started process=#{Process.pid} multiprocess=#{settings[:multiprocess]}")
+        ensure
+          lockfile.flock(File::LOCK_UN)
         end
+      end
 
-        @pid = Process.pid
+      def started?
+        Process.kill(0, pid)
+        true
+      rescue Errno::ESRCH, TypeError # process no longer exists or we had no PID cached
+        false
       end
-    end
 
-    def self.stop
-      # in puma clustered mode, we should only stop the daemon
-      # from the parent process (where it actually is running), not forked workers
-      return unless @pid == Process.pid
+      def start_services
+        services.values.each(&:run)
+      end
 
-      services.values.each(&:close)
+      def runnable?
+        return false if started? || !settings[:enabled] || ::Foreman.in_rake? || Rails.env.test?
 
-      @pid = nil
-    end
+        if defined?(Rails::Console)
+          Rails.logger.debug("Not starting Katello Event Daemon in the Rails Console")
+          return false
+        end
 
-    def self.started?
-      @pid.present?
-    end
+        if $PROGRAM_NAME.match(/spring/)
+          Rails.logger.debug("Spring is running in the environment; not starting Katello Event Daemon")
+          return false
+        end
 
-    def self.services
-      {
-        candlepin_events: ::Katello::CandlepinEventListener,
-        katello_events: ::Katello::EventMonitor::PollerThread
-      }
+        if Rails.env.production? && $PROGRAM_NAME.match(/dynflow/)
+          Rails.logger.debug("This appears to be the Dynflow process; not starting Katello Event Daemon")
+          return false
+        end
```

Comment:
This seems to allow it to run if the program name matches /dynflow/ as long as its not a production environment, which seems unnecessary.  (You could just remove the production? check), unless i'm missing some scenario?

Reply:
Good catch I guess dynflow could theoretically be started on its own in dev

Reply:
Fixed